include/onnxruntime/core/providers/providers.h

```diff
@@ -3,11 +3,16 @@
 
 #pragma once
 
+struct OrtSessionOptions;
+struct OrtLogger;
+
 namespace onnxruntime {
 class IExecutionProvider;
 
 struct IExecutionProviderFactory {
   virtual ~IExecutionProviderFactory() = default;
   virtual std::unique_ptr<IExecutionProvider> CreateProvider() = 0;
+  virtual std::unique_ptr<IExecutionProvider> CreateProvider(const OrtSessionOptions* session_options,
```

Comment:
Hi @skottmckay, FYI.

Here's the chain of reasoning that led me to add this version of CreateProvider():
- Need to be able to initialize `OrtModelCompilationOptions` from a const session options.
- The `OrtModelCompilationOptions` must create a copy of the original const session options.
- The original session options already has EPFactory instances added to it. These EPFactory instances always refer to the original session options, but need to be able to create EP instances with the _new_ session options with compile settings.
- Therefore, need to add a CreateProvider() factory method that takes in session_options. I believe this is aligned with what we discussed in your dev branch.

Reply:
WFM. I think all EPs in general should have access to the session options given we'll be doing more config via that, and they should all use the session logger.

Can we make the args references? I'd expect they're both required.

nit: logger -> session_logger so it's slightly clearer what it represents (i.e. it's not the default logger).

Reply:
I made both params const references and updated `logger` to `session_logger`.